apps/backend/src/opencage/opencage.controller.ts

```diff
@@ -0,0 +1,31 @@
+import { Controller, Get, Query } from '@nestjs/common';
+import * as https from 'https';
+
+@Controller('api/v1/location')
+export class OpencageController {
```

Comment:
i'd rather hide from our code which service we're using. i also think this controller is holding too much logic that it's not responsible for. WDYT about creating a `GeolocationService` interface and making a concrete implementation for opencage? then our code would only know this interface, not the implementation details. you can use dependency injection to provide the implementation.

Reply:
also, the opencage implementation should live under the `frameworks` folder

Reply:
I tried to create a GeolocationService interface and use the following snippet in the GeolocationModule providers:


``` typescript
providers: [
    { provide: GeolocationService, useClass: OpenCageService },
],
```
However, this did not work and resulted in the following error:

```
'GeolocationService' only refers to a type, but is being used as a value here.
```
Although I was unable to resolve this error, I was still able to refactor the code to improve it.

Reply:
so you probably tried to define`GeolocationService` as an interface right? by interface I meant something like an abstract class because by creating a class you can also use it as the key in the DI container. there's a `core/abstracts` folder where you can create something like this:

```js
interface ReversedGeocodingOutput {
  region: string;
  subregion: string;
}

export abstract class GeolocationService {
  public abstract reverseGeocoding(
    latitude: string,
    longitude: string,
  ): Promise<ReversedGeocodingOutput>;
}
```

then just extend it in the implementation:

```
@Injectable()
export class OpenCageService extends GeolocationService { }
```

it's important to mark the implementation as injectable so the container can inject the HTTP service later for you.